pygfx/controllers/_base.py

```diff
@@ -114,6 +280,308 @@ def add_default_event_handlers(self, *args):
     def update_camera(self, *args):
         raise DeprecationWarning("controller.update_camera() is no longer necessary")
 
+    # %% Builtin event handling
+
+    def handle_event(self, event, viewport):
+        if not self.enabled:
+            return
+        if not self._cameras:
+            return
+
+        rect = viewport.rect
+        need_update = False
+
+        type = event.type
+        if type.startswith(("pointer_", "key_", "wheel")):
+            modifiers = sorted([m.lower() for m in event.modifiers])
+            modifiers_prefix = "+".join(modifiers + [""])
+
+        if type == "before_render":
```

Comment:
should this also be conditional on `auto_update`?

Reply:
Yeah, that makes sense 👍